cmd/generate.go

```diff
@@ -1,21 +1,118 @@
 package cmd
 
 import (
+	"bytes"
+	"errors"
 	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
 
+	k8yaml "k8s.io/apimachinery/pkg/util/yaml"
+
+	kube "github.com/IBM/argocd-vault-plugin/pkg/kube"
 	"github.com/spf13/cobra"
 )
 
-// NewGenerateCommand Initializes the generate command
+// NewGenerateCommand initializes the generate command
 func NewGenerateCommand() *cobra.Command {
 	var command = &cobra.Command{
-		Use:   "generate",
-		Short: "Generate YAML files from template with Vault values",
+		Use:   "generate <path>",
+		Short: "Generate manifests from templates with Vault values",
 		RunE: func(cmd *cobra.Command, args []string) error {
-			fmt.Fprint(cmd.OutOrStdout(), "Generate YAML")
+
+			path := args[0]
+			files := listYamlFiles(path)
+			if len(files) < 1 {
+				return fmt.Errorf("no YAML files were found in %s", path)
+			}
+			manifests, errs := readFilesAsManifests(files)
+			if len(errs) != 0 {
+
+				// TODO: handle multiple errors nicely
+				return fmt.Errorf("could not read YAML files: %s", errs)
+			}
+
+			var resource kube.Template
+			for _, manifest := range manifests {
+				switch manifest["kind"] {
+				case "Deployment":
+					{
+						resource = kube.NewDeploymentTemplate(manifest)
+					}
+				case "Secret":
+					{
+						resource = kube.NewSecretTemplate(manifest)
+					}
+				}
+
+				err := resource.Replace()
+				if err != nil {
+					return err
+				}
+
+				output, err := resource.ToYAML()
+				if err != nil {
+					return err
+				}
+
+				fmt.Printf("%s---\n", output)
+			}
+
+			return nil
+		},
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) < 1 {
+				return errors.New("<path> argument required to generate manifests")
+			}
 			return nil
 		},
 	}
 
 	return command
 }
+
+func listYamlFiles(root string) []string {
+	var files []string
+
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if filepath.Ext(path) == ".yaml" {
+			files = append(files, path)
+		}
+		return nil
+	})
+	if err != nil {
+		panic(err)
```

Comment:
No `panic`, we want to make sure to return the error and log it  We always want to return a meaningful error

Reply:
Agreed about meaningful errors, but I guess I wonder what the purpose of `panic` actually is then? I thought it was for unrecoverable situations. Like why ever `panic` if you can just return `error`'s all the way back up?

Reply:
FYI there is still a `panic(error)` in `pkg/kube/util`'s `replaceInner` in the cases where we cannot convert to an inner YAML node to `map[string]interface{}`. Is that OK? I think I put that there b/c that was unexpected situation and thought panic was appropriate

Reply:
We will just need to see how it handles within Argo but it’s probably fine for now